clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2000,6 +2003,58 @@ protected void processCommonConsumerAction(
     }
   }
 
+  /**
+   * This function checks, for a regular user store, if its persisted offset is greater than the end offset of its
+   * corresponding Kafka topic (indicating an offset rewind event, thus a potential data loss in Kafka) and increases
+   * related sensor counter values.
+   */
+  private void reportUserStoreOffsetRewindMetrics(PartitionConsumptionState pcs) {
+    if (!isUserSystemStore()) {
+      return;
+    }
+
+    long offset = pcs.getLatestProcessedLocalVersionTopicOffset();
+    if (offset == OffsetRecord.LOWEST_OFFSET) {
+      return;
+    }
+    // Proceed if persisted OffsetRecord exists and has meaningful content.
+    long endOffset = getKafkaTopicPartitionEndOffSet(localKafkaServer, kafkaVersionTopic, pcs.getPartition());
+    if (endOffset != StatsErrorCode.LAG_MEASUREMENT_FAILURE.code && offset > endOffset) {
+      // report offset rewind.
+      LOGGER.warn(
+          "Offset rewind for version topic: {}, persisted record offset: {}, Kafka topic partition end-offset: {}",
+          kafkaVersionTopic,
+          offset,
+          endOffset);
+      versionedIngestionStats.recordIngestionOffsetRewind(storeName, versionNumber);
+    }
+  }
+
+  /**
+   * @return the end offset in kafka for the topic partition in SIT.
+   */
+  protected long getKafkaTopicPartitionEndOffSet(String kafkaUrl, String kafkaVersionTopic, int partition) {
+    long offsetFromConsumer = getPartitionLatestOffset(kafkaUrl, kafkaVersionTopic, partition);
+    if (offsetFromConsumer >= 0) {
```

Comment:
I saw you comparing against `OffsetRecord.LOWEST_OFFSET` earlier. Maybe we should try to use the same comparison if the semantic of the check is the same?

Reply:
I see, and there's no constant from Kafka side for the end offset of an empty topic?

Reply:
No, probably not. The normal way to identify an empty Kafka topic would be to check if its beginning offset == end offset.